### ISSUE TITLE
WeeklyPaper2. 머신러닝

### ISSUE THREAD

**Issue by finneKIM** [finneKIM/WeeklyPaper#2]
<html>
<body>
<h3>1. <strong>지도 학습(Supervised Learning)과 비지도 학습(Unsupervised Learning)의 차이</strong></h3>

항목 | 지도 학습 (Supervised) | 비지도 학습 (Unsupervised)
-- | -- | --
데이터 특징 | 입력값(X) + 정답(Y) 존재 | 입력값(X)만 존재, 정답 없음
목표 | 주어진 정답을 잘 맞추도록 학습 | 데이터 속 숨겨진 구조나 패턴을 찾는 것
예시 | 이메일 스팸 여부 분류, 집값 예측 등 | 고객군 분류(군집화), 차원 축소 등
대표 알고리즘 | 선형회귀, 로지스틱 회귀, SVM, 결정트리 등 | K-평균, PCA, DBSCAN 등
<html>
<body>
<h3>💡 쉽게 말하면:</h3>
<ul>
<li><strong>지도 학습</strong>: 선생님이 정답을 알려주고, 그 정답을 따라 배우는 방식</li>
<li><strong>비지도 학습</strong>: 선생님이 정답을 안 알려줌. 대신 스스로 패턴을 찾아야 함</li>
</ul>
<hr>
<h3>2. <strong>손실 함수(Loss Function)란 무엇이며, 왜 중요한가?</strong></h3>
<h3>✅ 손실 함수란?</h3>
<ul>
<li>머신러닝 모델이 <strong>얼마나 틀렸는지를 수치로 나타내는 함수</strong></li>
<li>예측값과 실제 정답 사이의 차이를 계산함</li>
</ul>
<h3>✅ 왜 중요한가?</h3>
<ol>
<li><strong>모델이 학습할 때 ‘방향’을 제시함</strong>
<ul>
<li>손실이 클수록 &quot;이 방향으로 가면 안 돼!&quot;라는 신호를 줌</li>
<li>손실이 작아질수록 모델이 정답에 가까워지고 있다는 뜻</li>
</ul>
</li>
<li>*경사 하강법(Gradient Descent)**에서 사용됨
<ul>
<li>손실 함수의 값을 줄이는 방향으로 **가중치(Weight)**를 조정</li>
</ul>
</li>
</ol>
<h3>✅ 종류 예시:</h3>

손실 함수 | 용도 | 설명
-- | -- | --
MSE (Mean Squared Error) | 회귀 | 오차 제곱 평균
Cross-Entropy | 분류 | 예측 확률과 실제값 차이를 측정

<h3>💡 비유:</h3>
<blockquote>
<p>손실 함수는 마치 시험 점수와 같고, 점수가 낮으면(손실이 큼) 공부 방향을 바꿔야 하고, 점수가 높아지면(손실이 작아짐) 잘하고 있다는 뜻.</p>
</blockquote>
<hr>

</body>
</html>

<h3>3. <strong>모델 학습 시 발생할 수 있는 편향(Bias)과 분산(Variance), 그리고 관계</strong></h3>

</body>
</html>
<html>
<body>
<h3>✅ 편향(Bias)이란?</h3>
<ul>
<li>모델이 <strong>너무 단순해서</strong> 정답을 잘 예측하지 못하는 경우</li>
<li>학습 데이터든 테스트 데이터든 <strong>다 틀리게 예측함</strong></li>
</ul>
<blockquote>
<p>예: 너무 단순한 모델로 복잡한 문제를 풀려 해서 틀리는 것</p>
<p>→ **언더피팅(underfitting)**이라고도 함</p>
</blockquote>
<h3>✅ 분산(Variance)이란?</h3>
<ul>
<li>모델이 <strong>너무 복잡해서</strong> 학습 데이터에는 잘 맞지만, <strong>새로운 데이터에는 잘 틀리는</strong> 경우</li>
</ul>
<blockquote>
<p>예: 너무 외운 듯한 모델, 시험 공부한 문제는 맞지만 다른 문제는 다 틀림</p>
<p>→ **오버피팅(overfitting)**이라고 함</p>
</blockquote>
<h3>✅ 관계:</h3>
<ul>
<li><strong>편향과 분산은 서로 트레이드오프(trade-off)</strong> 관계
<ul>
<li>너무 단순하면 → 편향 ↑, 분산 ↓</li>
<li>너무 복잡하면 → 편향 ↓, 분산 ↑</li>
</ul>
</li>
</ul>
<blockquote>
<p>목표는 이 두 가지를 균형 있게 조절해서 일반화 성능을 높이는 것</p>
</blockquote>
<hr>
<h3>4. <strong>K-폴드 교차 검증(K-Fold Cross Validation)에서 K 값 선택 시 고려할 점</strong></h3>
<h3>✅ K-폴드란?</h3>
<ul>
<li>데이터를 K개로 나누어, 그 중 1개는 <strong>검증용</strong>, 나머지는 <strong>훈련용</strong>으로 사용</li>
<li>이 과정을 K번 반복하여 <strong>평균 성능을 평가</strong></li>
</ul>
<h3>✅ K 값 선택 시 고려 사항:</h3>
<ol>
<li><strong>데이터 크기</strong>
<ul>
<li>데이터가 작으면 K를 크게 (예: 10)</li>
<li>데이터가 크면 K를 작게 해도 괜찮음 (예: 5)</li>
</ul>
</li>
<li><strong>계산 비용</strong>
<ul>
<li>K가 클수록 훈련을 많이 하므로 <strong>시간과 자원이 더 듬</strong></li>
<li>예: K=10이면 모델을 10번 학습해야 함</li>
</ul>
</li>
<li><strong>편향 vs 분산의 균형</strong>
<ul>
<li>K가 크면 검증 데이터는 작아지고 훈련 데이터는 많아져 → <strong>분산이 낮아지고 편향이 낮아짐</strong></li>
<li>K가 작으면 검증 데이터는 커지고 훈련 데이터는 작아져 → <strong>평가가 덜 안정적</strong></li>
</ul>
</li>
</ol>
<h3>✅ 일반적인 K 값 추천:</h3>

데이터 양 | 추천 K 값
-- | --
작을 때 | 10
중간 | 5~10
클 때 | 5 또는 Hold-out 방식(단순 분할)도 가능

</body>
</html>

**Comment by finneKIM** [comment]
<html>
<body>
<h2>✅ 1. 지도 학습 vs 비지도 학습</h2>
<h3>📊 시각화</h3>
<pre><code>
지도 학습
(X, Y) 라벨이 있음
-------------------------
입력: [사진] → 출력: [고양이]
입력: [키, 몸무게] → 출력: [비만 여부]

비지도 학습
(X만 있음) 라벨 없음
-------------------------
입력: [고객 구매 이력] → 자동으로 비슷한 고객 그룹으로 나눔
입력: [뉴스 기사] → 주제별로 군집화

</code></pre>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/e0/Comparison_machine_learning_paradigms.svg/800px-Comparison_machine_learning_paradigms.svg.png" alt=""></p>
<blockquote>
<p>⬅ 지도학습은 정답(Y)이 있고</p>
<p>➡ 비지도학습은 정답 없이 패턴만 찾음</p>
</blockquote>
<hr>
<h3>🧪 예제 코드 (Python, scikit-learn 기반)</h3>
<h3>1) 지도학습 - 의사결정나무 분류기</h3>
<pre><code class="language-python">
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# 꽃 데이터셋 불러오기
X, y = load_iris(return_X_y=True)

# 데이터 나누기 (훈련:테스트 = 80:20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# 분류기 훈련
model = DecisionTreeClassifier()
model.fit(X_train, y_train)

print(&quot;정확도:&quot;, model.score(X_test, y_test))

</code></pre>
<h3>2) 비지도학습 - KMeans 클러스터링</h3>
<pre><code class="language-python">
from sklearn.datasets import load_iris
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

iris = load_iris()
X = iris.data

kmeans = KMeans(n_clusters=3, random_state=42)
y_kmeans = kmeans.fit_predict(X)

plt.scatter(X[:, 0], X[:, 1], c=y_kmeans)
plt.title(&quot;비지도학습 (KMeans 클러스터링)&quot;)
plt.xlabel(&quot;꽃받침 길이&quot;)
plt.ylabel(&quot;꽃받침 너비&quot;)
plt.show()

</code></pre>
<hr>
<h2>✅ 2. 손실 함수(Loss Function)</h2>
<h3>📊 시각화</h3>
<h3>MSE (Mean Squared Error)</h3>
<ul>
<li>예측값과 실제값의 차이의 <strong>제곱 평균</strong></li>
<li>값이 클수록 오차가 크다는 뜻</li>
</ul>
<pre><code>
실제값:      ●
예측값1:    ●       ← 손실 작음
예측값2:         ● ← 손실 큼

손실(Loss) = (예측값 - 실제값)^2

</code></pre>
<hr>
<h3>🧪 예제 코드: MSE vs MAE 비교</h3>
<pre><code class="language-python">
from sklearn.metrics import mean_squared_error, mean_absolute_error

y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]

# 평균제곱오차
mse = mean_squared_error(y_true, y_pred)
# 평균절대오차
mae = mean_absolute_error(y_true, y_pred)

print(&quot;MSE:&quot;, mse)
print(&quot;MAE:&quot;, mae)

</code></pre>
<hr>
<h2>✅ 3. 편향(Bias)과 분산(Variance)</h2>
<h3>📊 시각화</h3>

유형 | 특징 | 문제
-- | -- | --
편향 ↑ | 단순 모델 | 모든 데이터에서 오답 (언더피팅)
분산 ↑ | 너무 복잡 | 훈련 데이터에만 잘 맞춤 (오버피팅)


<blockquote>
<p>🎯 목표는 편향과 분산을 모두 적당히 줄이는 것</p>
</blockquote>
<hr>
<h3>🧪 예제 코드: 고차 다항 회귀에서 오버피팅/언더피팅 확인</h3>
<pre><code class="language-python">
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# 예제 데이터 생성
np.random.seed(0)
X = np.sort(np.random.rand(40, 1) * 2 - 1, axis=0)
y = X**3 + 0.1 * np.random.randn(40, 1)

# 다양한 차수로 모델 학습
degrees = [1, 3, 10]
plt.figure(figsize=(12, 4))

for i, degree in enumerate(degrees):
    poly = PolynomialFeatures(degree)
    X_poly = poly.fit_transform(X)

    model = LinearRegression().fit(X_poly, y)
    y_pred = model.predict(X_poly)

    plt.subplot(1, 3, i + 1)
    plt.scatter(X, y, color='black')
    plt.plot(X, y_pred, color='blue')
    plt.title(f'{degree}차 모델')
plt.show()

</code></pre>
<ul>
<li><strong>1차</strong>: 편향 ↑</li>
<li><strong>10차</strong>: 분산 ↑</li>
<li><strong>3차</strong>: 적절</li>
</ul>
<hr>
<h2>✅ 4. K-폴드 교차 검증</h2>
<h3>📊 시각화</h3>
<pre><code>
K=5일 때 예시
Fold1: [검증]      [훈련훈련훈련훈련]
Fold2: [훈련] [검증]    [훈련훈련훈련]
Fold3: [훈련훈련] [검증]   [훈련훈련]
Fold4: [훈련훈련훈련] [검증]  [훈련]
Fold5: [훈련훈련훈련훈련] [검증]

</code></pre>
<ul>
<li>모델을 5번 훈련시키고, 성능을 평균 내어 평가의 <strong>신뢰성 증가</strong></li>
</ul>
<hr>
<h3>🧪 예제 코드: K-Fold 교차 검증</h3>
<pre><code class="language-python">
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold, cross_val_score

X, y = load_iris(return_X_y=True)
model = LogisticRegression(max_iter=200)

# K-Fold 설정
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# 교차 검증 수행
scores = cross_val_score(model, X, y, cv=kf)
print(&quot;각 Fold별 정확도:&quot;, scores)
print(&quot;평균 정확도:&quot;, scores.mean())

</code></pre>

</body>
</html>

**Comment by finneKIM** [comment]
<html>
<body>
<h1>산과 편향, 관계에 대한 코멘트 추가</h1>
<h2>✅ 1. 편향과 분산이란 무엇인가?</h2>

개념 | 설명 | 비유
-- | -- | --
편향 (Bias) | 예측값의 평균과 실제값 사이의 차이 | "자꾸 틀리는 방향이 같아"
(모델이 잘못된 가정에 묶임) |   |  
분산 (Variance) | 데이터가 바뀔 때 예측값의 변동성 | "문제 바뀌면 틀려버려"
(너무 민감하게 반응함) |   |  


<hr>
<h2>✅ 2. 수학적 정의</h2>
<blockquote>
<p>아래는 회귀 문제에서 가장 흔히 쓰이는 수식(참고)</p>
</blockquote>
<h3>🎯 실제 값: <code>y</code></h3>
<h3>🎯 예측 함수: <code>f̂(x)</code> (모델이 예측한 값)</h3>
<h3>✅ (1) <strong>편향</strong> 계산 방법</h3>
<pre><code>Bias(x)^2 = ( E[f̂(x)] - f(x) )^2

</code></pre>
<ul>
<li><code>E[f̂(x)]</code>: 여러 번 학습했을 때 <strong>평균 예측값</strong></li>
<li><code>f(x)</code>: 실제 목표 함수 (현실의 정답)</li>
</ul>
<p>📌 즉, 편향은 <strong>예측 평균이 실제 정답에서 얼마나 먼가</strong>를 보는 것</p>
<hr>
<h3>✅ (2) <strong>분산</strong> 계산 방법</h3>
<pre><code>Variance(x) = E[ (f̂(x) - E[f̂(x)])^2 ]

</code></pre>
<ul>
<li>즉, 데이터가 조금만 바뀌어도 모델 예측값이 <strong>얼마나 흔들리는가</strong>를 측정</li>
</ul>
<hr>
<h3>✅ (3) 총 오차: MSE로 표현</h3>
<pre><code>MSE(x) = Bias(x)^2 + Variance(x) + Irreducible Error

</code></pre>
<ul>
<li><strong>Irreducible Error</strong>: 우리가 어쩔 수 없는 자연스러운 오차 (ex. 센서 오류, 측정 오차)</li>
</ul>

</body>
</html>

<html>
<body>
<ul>
<li><strong>Irreducible Error</strong>: 우리가 어쩔 수 없는 자연스러운 오차 (ex. 센서 오류, 측정 오차)</li>
</ul>
<hr>
<h2>✅ 3. 왜 복잡한 모델은 분산이 크고, 단순한 모델은 편향이 큰가?</h2>
<h3>📌 복잡한 모델 (예: 10차 다항식)</h3>
<ul>
<li>데이터의 작은 패턴까지 전부 외우려고 함</li>
<li>훈련 데이터에는 잘 맞지만, <strong>새로운 데이터에는 일반화 못함</strong></li>
<li>따라서 <strong>분산(Variance)↑</strong>, <strong>편향(Bias)↓</strong></li>
</ul>
<h3>📌 단순한 모델 (예: 직선)</h3>
<ul>
<li>현실은 복잡한데 너무 간단하게 보려 함</li>
<li>모든 데이터에 대해 <strong>비슷하게 틀림</strong></li>
<li>따라서 <strong>편향(Bias)↑</strong>, <strong>분산(Variance)↓</strong></li>
</ul>
<hr>
<h2>✅ 4. 시각적 이해 (타겟 예시)</h2>
<pre><code>🎯🎯🎯 타겟판이 실제 정답이라고 해보자

편향 ↑
 -&gt; 화살이 한쪽에 몰려서 벗어남 (일관되게 틀림)

분산 ↑
 -&gt; 화살들이 여기저기 흩어짐 (불안정함)

편향↓ &amp; 분산↓
 -&gt; 정중앙에 모여있음 (완벽한 모델)

편향↑ &amp; 분산↑
 -&gt; 중앙도 아니고, 퍼지기까지 함 (최악)

</code></pre>
<hr>
<h2>✅ 5. 예제 코드 (시뮬레이션 기반)</h2>
<pre><code class="language-python">import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error

np.random.seed(0)

# 실제 함수 (정답)
def true_function(x):
    return np.sin(np.pi * x)

# 데이터 생성 함수
def create_dataset(n_samples=25):
    X = np.random.uniform(0, 1, size=n_samples)
    y = true_function(X) + np.random.normal(0, 0.1, size=n_samples)  # 노이즈 포함
    return X.reshape(-1, 1), y

# 편향-분산 계산용 시뮬레이션
def bias_variance_demo(degree=1, n_trials=100):
    x_plot = np.linspace(0, 1, 100).reshape(-1, 1)
    predictions = []

    for _ in range(n_trials):
        X, y = create_dataset()
        poly = PolynomialFeatures(degree=degree)
        X_poly = poly.fit_transform(X)
        x_plot_poly = poly.transform(x_plot)
        model = LinearRegression().fit(X_poly, y)
        y_pred = model.predict(x_plot_poly)
        predictions.append(y_pred)

    predictions = np.array(predictions)
    avg_prediction = np.mean(predictions, axis=0)
    true_y = true_function(x_plot.flatten())

    # 편향 계산
    bias_squared = np.mean((avg_prediction - true_y) ** 2)
    # 분산 계산
    variance = np.mean(np.var(predictions, axis=0))
    # 전체 오차
    mse = bias_squared + variance

    return bias_squared, variance, mse

# 다양한 모델 복잡도에 대한 시뮬레이션
for deg in [1, 3, 9]:
    bias2, var, mse = bias_variance_demo(degree=deg)
    print(f&quot;[{deg}차 다항식] Bias^2: {bias2:.4f}, Variance: {var:.4f}, MSE: {mse:.4f}&quot;)

</code></pre>
<hr>
<h2>✅ 요약</h2>

구분 | 설명 | 높을 때 문제
-- | -- | --
편향(Bias) | 모델의 단순함 → 정답과 평균 예측의 거리 | 언더피팅
분산(Variance) | 모델의 민감함 → 데이터 바뀌면 예측이 출렁 | 오버피팅



</body>
</html>
